zanata-war/src/main/java/org/zanata/action/MergeTransAction.java

```diff
@@ -0,0 +1,230 @@
+    }
+
+    public List<HProjectIteration> getSourceVersions() {
+        List<HProjectIteration> versions =
+                getSourceProject().getProjectIterations();
+
+        if(versions.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        //remove target version if both are the same project or obsolete
+        if (StringUtils.equals(sourceProjectSlug, targetProjectSlug)) {
+            for (HProjectIteration version : versions) {
+                if (version.getSlug().equals(targetVersionSlug)
+                        || version.getStatus().equals(EntityStatus.OBSOLETE)) {
+                    versions.remove(version);
+                    break;
+                }
+            }
+        }
+        return versions;
+    }
+
+    public List<HProject> getProjects() {
+        return projectDAO.getProjects(EntityStatus.ACTIVE,
+                EntityStatus.READONLY);
+    }
+
+    public void startMergeTranslations() {
+        if (StringUtils.isEmpty(sourceProjectSlug)
+                || StringUtils.isEmpty(sourceVersionSlug)
+                || StringUtils.isEmpty(targetProjectSlug)
+                || StringUtils.isEmpty(targetVersionSlug)) {
+            FacesMessages.instance().add(StatusMessage.Severity.WARN,
+                msgs.get("jsf.iteration.mergeTrans.noSourceAndTarget"));
+            return;
+        }
+        if (isCopyActionsRunning()) {
+            FacesMessages.instance().add(StatusMessage.Severity.WARN,
+                msgs.get("jsf.iteration.mergeTrans.hasCopyActionRunning"));
+            return;
+        }
+        mergeTranslationsManager.startMergeTranslations(sourceProjectSlug,
+                sourceVersionSlug, targetProjectSlug, targetVersionSlug,
+                !keepExistingTranslation);
+    }
+
+    // Check if copy-trans, copy version or merge-trans is running for given
+    // version
+    public boolean isCopyActionsRunning() {
+        return mergeTranslationsManager.isMergeTranslationsRunning(
+                targetProjectSlug, targetVersionSlug)
+                || copyVersionManager.isCopyVersionRunning(targetProjectSlug,
+                        targetVersionSlug) ||
+                copyTransManager.isCopyTransRunning(getTargetVersion());
+    }
+
+    @Override
+    public boolean isInProgress() {
+        return mergeTranslationsManager.isMergeTranslationsRunning(
+            targetProjectSlug, targetVersionSlug);
+    }
+
+    @Override
+    public String getCompletedPercentage() {
+        MergeTranslationsTaskHandle handle = getHandle();
+        if (handle != null) {
+            double completedPercent =
+                (double) handle.getCurrentProgress() / (double) handle
+                    .getMaxProgress() * 100;
+            if (Double.compare(completedPercent, 100) == 0) {
+                onComplete();
+            }
+            return PERCENT_FORMAT.format(completedPercent);
+        } else {
+            return "0";
+        }
+    }
```

Comment:
It looks like this code has been copied around several times. It should go in a utility class, or maybe pulled up into the task handle class.